### ISSUE TITLE
A design for accumulators in the core IR

### ISSUE THREAD

**Issue by athas** [diku-dk/futhark#1051]
We have recently realised the need to express histogram-like
computations where each input element can give rise to an arbitrary
number of updates.  #737 lists examples (but that can be worked around
with padding or `expand`), and our recent AD research revealed more
cases that cannot simply be worked around.  Therefore, I think we
should start designing and implementing the **core IR** representation
of such a feature.  Let me repeat: this issue is **only concerned with
the IR**, *not* the source language.  This means we do not need
guaranteed soundness, but we need to have a design and rules
(preferably checkable) that make it easy for the compiler to do the
right thing.  How we expose this to the source language (if at all) is
for later.

Since the core IR does not have a well-defined notation for type
rules, I am going to approximate some of the rules through source
language syntax.  **This is still only about IR constructs.**

The basic idea that we have come up with is to introduce a type of
*accumulators*, by which we can *write* values, but not read.
Accumulators are constructed from arrays and can also be turned back
into arrays:

```
type acc [n] 't
val mkacc [n] 't : (t -> t -> t) -> *[n]t -> *(acc [n] t)
val upd [n] 't : *(acc [n] t) -> i32 -> t -> *(acc [n] t)
val unacc [n] 't : *(acc [n] t) -> [n]t
```

When you convert an array to an accumulator with `mkacc`, you also
provide an associative and commutative binary operator, which is then
used whenever `upd` is used to update a value at a given index.  This
means that the order of `upd`s does not matter.  Multiple threads can
call `upd` on the same `acc` concurrently, without risk.  However,
*how* do we write parallel updates of a single `acc`?  We could try
writing something like this:

```
let xs_acc = mkacc (+) xs
let ys = map (\i -> upd xs_acc i 1) is
```

But this is not OK.  The `upd` function consumes its result, to
enforce linear usage of the accumulator, and all values become
non-unique inside of a lambda.  We need this instead:

```
let xs_acc = mkacc (+) xs
let xs_accs = replicate n xs_acc
let ys = map (\i -> upd xs_acc i 1) is xs_accs
```

But we still have a problem - that `ys` is an array of *accumulators*.
Each of those can be turned into an array with `unacc`, but that is
not really what we want.  We need a new primitive for collapsing
accumulators:

```
val join_accs [n] t : *[](acc [n] t) -> *(acc [n] t)
```

Alternatively, `unacc` should accept an array of accumulators instead.
Maybe `mkacc` should also produce an array, since accumulators are
*only* worth using over ordinary in-place updates when you have a
parallel operation.

But there is another problem.  This design is (sort of) sound because
we cannot *read* from an accumulator, only write to it.  It is
therefore important that we do not ever have an array coexist with an
accumulator for that array.  This is partially handled by `mkacc`
consuming the input array, but the ability to copy accumulators gives
us problems:

```
let acc = mkacc (+) xs
let acc2 = copy acc
let xs' = unacc acc
```

It is not feasible to simply ban copying of accumulators, since
returning an accumulator from `map` is exactly the same as a copy, and
we *need* that.  This is also bad:

```
let xs_acc = mkacc (+) xs
let xs_accs = replicate n xs_acc
let ys = map (\i -> upd xs_acc i 1) is xs_accs
let ys1 = unacc ys[0]
let ys2 = unacc ys[1]
```

To address this, perhaps we can inject the name of the array covered
by an accumulator *into the accumulator type itself*.  There is no way
to express this in pseudo-source-language-syntax, but I'm thinking
something like this:

```
let (xs_acc : acc xs) = mkacc (+) xs
let (xs_accs : [n](acc xs)) = replicate n xs_acc
```

This brings two type rules:

1. An accumulator *cannot* escape the scope of its corresponding array.

2. The `unacc` operation can only be done *once* for any array
   accessed through accumulator.

There is no way these rules would fly in the source language, but the
IR is much easier to extend in ad hoc ways.

**Comment by athas** [comment]
I've worked on adding accumulators to the core language.  That's still in-progress.  For testing, accumulators must also be exposed somehow in the source language.  I've come up with an interface that will hopefully be sound, and it looks like this:

```
type~ acc 't = intrinsics.acc t

let scatter_stream [k] 'a 'b
                   (dest: *[k]a)
                   (f: *(acc a) -> b -> *(acc a))
                   (bs: []b)
                 : *[k]a =
  intrinsics.scatter_stream (dest, f, bs) :> [k]a

let reduce_by_index_stream [k] 'a 'b
                   (dest: *[k]a)
                   (op: a -> a -> a)
                   (ne: a)
                   (f: *(acc a) -> b -> *(acc a))
                   (bs: []b)
                 : *[k]a =
  intrinsics.hist_stream (dest, op, ne, f, bs) :> [k]a

let write 't (acc : *(acc t)) (i: i32) (v: t) : *(acc t) =
  intrinsics.acc_write (acc, i, v)

let main (xs: *[]i32) =
  let f (acc: *(acc i32)) i =
    let acc = write acc (i*2) i
    let acc = write () (i*2+1) i
    in acc
  in reduce_by_index_stream xs (+) 0 f (iota 10)

```

I think my next step will be to implement accumulators in the interpreter, so we can be sure this stuff actually has reasonable semantics.

**Comment by athas** [comment]
I have this source-language API working in the interpreter:

```futhark
type~ acc t -- abstract

val scatter_stream [k] a b [impl₀] : (dest: *[k]a) -> (f: *acca -> b -> *acc a) -> (bs: [impl₀]b) -> *[k]a

val reduce_by_index_stream [k] a b [impl₀] : (dest: *[k]a) -> (op: a -> a -> a) -> (ne: a) -> (f: *acc a -> b -> *acc a) -> (bs: [impl₀]b) -> *[k]a

val write t : (acc: *acc t) -> (i: i32) -> (v: t) -> *acc t
```

**Comment by athas** [comment]
The *really difficult* problem has been overcome.  Things work in the sequential backends.  Now the challenge is to make it work in the GPU backends.

**Comment by athas** [comment]
Closed in #1197.
